### PR TITLE
[WIP][SYCL] Add support for union types kernel arguments

### PR DESCRIPTION
**Pull request by smanna12**
This patch adds support for union types as kernel parameter using existing codes of 
"handleScalarType" and renames "handleScalarType"  to "handleSimpleType" to 
be more generic for scalars, vectors and unions types. 

 -Add kernel generation, integration header test, and runtime tests.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

**Comment by smanna12** [comment]
Patch is incomplete. Work in progress.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -941,9 +943,11 @@ class KernelObjVisitor {
       else if (FieldTy->isArrayType()) {
         if (KF_FOR_EACH(handleArrayType, Field, FieldTy))
           VisitArrayElements(Field, FieldTy, handlers...);
-      } else if (FieldTy->isScalarType() || FieldTy->isVectorType())
+      } else if (FieldTy->isScalarType() || FieldTy->isVectorType()) {
         KF_FOR_EACH(handleScalarType, Field, FieldTy);
-      else
+      } else if (FieldTy->isUnionType()) {
+        KF_FOR_EACH(handleUnionType, Field, FieldTy);
+      } else
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:950]: If code of "handleScalarType" in "IntHeaderCreator" also suits for unions, we could unify it across scalars, unions and vectors, i.e. here you will have something like
```
} else if (FieldTy->isScalarType() || FieldTy->isVectorType()
 || FieldTy->isUnionType()) {
KF_FOR_EACH(handleSimpleType, Field, FieldTy);
}
```
i.e. `handleScalarType` can be renamed to something more generic (`handleSimpleType` in my example) and used for scalars, vectors and unions..

Comment (smanna12) [reply clang/lib/Sema/SemaSYCL.cpp]: @Fznamznon, It seems like"handleScalarType" suits for unions. So this is renamed to "handleSimpleType".

**Comment by smanna12** [comment]
Sorry for the several submissions. Clang-format seems like picking wrong patches. Failing.

**Comment by smanna12** [comment]
Please do not review. Sorry wrong merge conflicts happen.

**Comment by smanna12** [comment]
I have created separate PR here: https://github.com/intel/llvm/pull/2255

**Pull request closed without merging**